src/runtime/define.ts

```diff
@@ -31,61 +32,101 @@ export type Definition = {
   assets?: Map<string, string>;
 };
 
-export function define(state: DefineState, definition: Definition, observer = observe): void {
-  const {id, body, inputs = [], outputs = [], output, autodisplay, autoview, automutable} = definition;
-  const variables = state.variables;
-  const v = main.variable(observer(state, definition), {shadow: {}});
-  const vid = output ?? (outputs.length ? `cell ${id}` : null);
-  if (inputs.includes("display") || inputs.includes("view")) {
-    let displayVersion = -1; // the variable._version of currently-displayed values
-    const vd = new (v.constructor as typeof Variable)(2, v._module);
-    vd.define(
-      inputs.filter((i) => i !== "display" && i !== "view"),
-      () => {
-        const version = v._version; // capture version on input change
-        return (value: unknown) => {
-          if (version < displayVersion) throw new Error("stale display");
-          else if (state.variables[0] !== v) throw new Error("stale display");
-          else if (version > displayVersion) clear(state);
-          displayVersion = version;
-          display(state, value);
-          return value;
-        };
-      }
-    );
-    v._shadow.set("display", vd);
-    if (inputs.includes("view")) {
-      const vv = new (v.constructor as typeof Variable)(2, v._module, null, {shadow: {}});
-      vv._shadow.set("display", vd);
-      vv.define(["display"], (display) => (value: unknown) => input(display(value)));
-      v._shadow.set("view", vv);
+export class NotebookInstance {
+
+  protected runtime = Object.assign(new Runtime({ ...library, __ojs_runtime: () => this.runtime }), { fileAttachments });
+  protected main = (this.runtime as typeof this.runtime & { main: Module }).main = this.runtime.module();
+  protected stateById = new Map<number, DefineState>();
+
+  constructor() {
+    this.main.constructor.prototype.defines = function (this: Module, name: string): boolean {
+      return (
+        this._scope.has(name) ||
+        this._builtins.has(name) ||
+        this._runtime._builtin._scope.has(name)
+      );
+    };
```

Comment:
With the `defaultInstance` approach, this mutation could be done statically. (But really we should move it to the Runtime itself… Future work!)